File: CodingInterviews/Merge.java
public class Merge {
    public static void main(String[] args) {
        int[] A = { 1, 2, 3, 0, 0, 0 };
        int[] B = { 2, 5, 6 };
        merge(A, 3, B, 3);
        for (var i : A) {
            System.out.print(i);
            System.out.print('\t');
        }
    }

    private static void merge(int[] A, int m, int[] B, int n) {
        int pa = m - 1;
        int pb = n - 1;
        int i = m + n - 1;
        // merge
        while (pa >= 0 && pb >= 0)
            A[i--] = A[pa] > B[pb] ? A[pa--] : B[pb--];
        // merge rest
        while (pa >= 0)
            A[i--] = A[pa--];
        while (pb >= 0)
            A[i--] = B[pb--];
    }
}

File: CodingInterviews/NowCoder/Merge.cpp
#include "main.h"

class Solution {
  public:
    ListNode *Merge(ListNode *pHead1, ListNode *pHead2) {
        ListNode head(-1), *p = &head, *t;
        while (pHead1 && pHead2) {
            if (pHead1->val <= pHead2->val) {
                t = pHead1;
                pHead1 = pHead1->next;
            } else {
                t = pHead2;
                pHead2 = pHead2->next;
            }
            t->next = p->next;
            p->next = t;
            p = p->next;
        }
        p->next = pHead1 ? pHead1 : pHead2;
        return head.next;
    }
};